### PR TITLE
Adding MariaDB support to C++ Dev Container

### PR DESCRIPTION
**Pull request by KanishkT123**
# [Work Item ID](https://github.com/microsoft/vscode-dev-containers/issues/920)

## Description

---

This PR adds:
1. New Database and Language pair for C++ and MariaDB. 
2. Tests to verify that the Database is running and accepts commands from the C++ container.
3. Supporting documentation for running and editing the containers locally

## PR Checklist

---

> Use the check-list below to ensure your branch is ready for PR.  If the item is not applicable, leave it blank.

- [X] I have updated the documentation accordingly.
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.
- [X] My code follows the code style of this project.
- [X] I ran the lint checks which produced no new errors nor warnings for my changes.
- [X] I have checked to ensure there aren't other open Pull Requests for the same update/change.

## Does this introduce a breaking change?

---

- [ ] Yes
- [X] No

## Testing

---

> - Instructions for testing and validation of your code:

All testing can be carried out using the existing test runner, `test.sh`. The README.md also contains explicit testing instructions.

>   - What OS was used for testing.
Windows 10, Windows 11. 
>   - Which test sets were used.
N/A
>   - Description of test scenarios that you have tried.
Tested building and compilation, setting up database, connecting to DB Container, logging into DB, sending and receiving data from DB.

## Other information or known dependencies

---

These changes are primarily additive, taken from the existing C++ devcontainer.

### CODE REVIEW

File: containers/cpp-mariadb/test-project/test.sh
Comment (juzuluag) [inline containers/cpp-mariadb/test-project/test.sh:13]: Looks like `main.out` is not executed after build. This way db connectivity would be checked.
```sh
vscode ➜ /workspace/test-project $ ./main.out 
Hello, Remote World!
DB Connecting
DB Executing
Cluster has the following user created databases
mariadb
DB Success
```

File: containers/cpp-mariadb/.devcontainer/devcontainer.json
Comment (juzuluag) [inline containers/cpp-mariadb/.devcontainer/devcontainer.json:13]: How about adding : `ms-vscode.cpptools-extension-pack` . Seems like useful for debugging.

Comment (KanishkT123) [reply containers/cpp-mariadb/.devcontainer/devcontainer.json]: Good call, thank you!

File: containers/cpp-mariadb/.devcontainer/devcontainer.json
Comment (juzuluag) [inline containers/cpp-mariadb/.devcontainer/devcontainer.json]: nit: Seems like spaces are not correct.

File: containers/cpp-mariadb/.devcontainer/install-mariadb.sh
Comment (juzuluag) [inline containers/cpp-mariadb/.devcontainer/install-mariadb.sh:47]: Shouldn't be here `exit 1`

File: containers/cpp-mariadb/.devcontainer/install-mariadb.sh
Comment (juzuluag) [inline containers/cpp-mariadb/.devcontainer/install-mariadb.sh:2]: Suggestion: Use bash and set to fail in case any command fails.

```bash
#!/bin/bash
set -e
```

File: containers/cpp-mariadb/.devcontainer/install-mariadb.sh
Comment (juzuluag) [inline containers/cpp-mariadb/.devcontainer/install-mariadb.sh:53]: Make sure to delete the downloaded assets.

File: containers/cpp-mariadb/.devcontainer/Dockerfile
Comment (juzuluag) [inline containers/cpp-mariadb/.devcontainer/Dockerfile:11]: If you create the file with +x, no need to `chmod` it as part of the `RUN` call

File: containers/cpp-mariadb/.devcontainer/devcontainer.json
Comment (dnastrain) [reply containers/cpp-mariadb/.devcontainer/devcontainer.json]: fixed

File: containers/cpp-mariadb/.devcontainer/Dockerfile
Comment (KanishkT123) [reply containers/cpp-mariadb/.devcontainer/Dockerfile:11]: Docker best practices seems to be to not make the file `+x` on creation but on run.

File: containers/cpp-mariadb/.devcontainer/install-mariadb.sh
Comment (KanishkT123) [reply containers/cpp-mariadb/.devcontainer/install-mariadb.sh:47]: Yes, sorry. Typo!

**Review by dnastrain — APPROVED** [review]
overall LGTM, after the recent edits we've made (an any in progress from recent comments). we'll await squads / leads to look it over.

File: containers/cpp-mariadb/.devcontainer/install-mariadb.sh
Comment (juzuluag) [reply containers/cpp-mariadb/.devcontainer/install-mariadb.sh:53]: Hi @KanishkT123 , I would recommend to use bash `trap` statement to clean up.

**Review by juzuluag — APPROVED** [review]
Other than the comments that need to be resolved, LGTM. Great work.

File: containers/cpp-mariadb/.devcontainer/install-mariadb.sh
Comment (juzuluag) [reply containers/cpp-mariadb/.devcontainer/install-mariadb.sh]: Implemented. https://github.com/microsoft/vscode-dev-containers/pull/1241/commits/a4744e7f76737926658e95a97db2b97f8eb4d713

File: containers/cpp-mariadb/README.md
Comment (juzuluag) [inline containers/cpp-mariadb/README.md:85]:
```suggestion
4. Select the `containers/cpp-mariadb` folder.
```

File: containers/cpp-mariadb/.devcontainer/install-mariadb.sh
Comment (Kpopovs1) [inline containers/cpp-mariadb/.devcontainer/install-mariadb.sh:10]: you don't need to say `function` to define the func name in this case, you can just do:

`cleanup() { ... }`

File: containers/cpp-mariadb/.devcontainer/install-mariadb.sh
Comment (Kpopovs1) [inline containers/cpp-mariadb/.devcontainer/install-mariadb.sh:15]: Since you are executing this cleanup function on exit, it may be worth:
- Returning the exit code
- adding a `+e` before removing the tmp dir so this doesn't fail if there is an issue

For example:

```sh
cleanup( ) {
    EXIT_CODE=$?
    set +e
    if [[ -n "${TMP_DIR}" ]]; then
         # cd /
         # remove all the tmp dir things
    fi
    exit $EXIT_CODE
}
trap cleanup EXIT
```

File: containers/cpp-mariadb/.devcontainer/install-mariadb.sh
Comment (Kpopovs1) [inline containers/cpp-mariadb/.devcontainer/install-mariadb.sh:28]: This `case` statement (lines 28 to 62) could potentially be extracted into a function that exports `OSURL` and
`OSTAG` to make the code more readable and cleaner.

File: containers/cpp-mariadb/test-project/main.cpp
Comment (Kpopovs1) [inline containers/cpp-mariadb/test-project/main.cpp:15]: A small nit for consistency with `{` usage throughout the cpp file: 

```cpp
function() {
    ...
}
```
 or

```cpp
function()
{
    ...
}
```

**Review by Kpopovs1 — COMMENTED** [review]
Looks great!! A few comments from my end.

File: containers/cpp-mariadb/.devcontainer/Dockerfile
Comment (xtophs) [inline containers/cpp-mariadb/.devcontainer/Dockerfile:5]: nit: lower case B in below?

File: containers/cpp-mariadb/.devcontainer/docker-compose.yml
Comment (xtophs) [inline containers/cpp-mariadb/.devcontainer/docker-compose.yml:43]: stable version tag?

File: containers/cpp-mariadb/.devcontainer/install-mariadb.sh
Comment (xtophs) [inline containers/cpp-mariadb/.devcontainer/install-mariadb.sh:1]: Did we write this or do we need attribution and license reference here?

File: containers/cpp-mariadb/.vscode/launch.json
Comment (xtophs) [inline containers/cpp-mariadb/.vscode/launch.json:2]: not valid json ... hence the red syntax highlighting

File: containers/cpp-mariadb/.vscode/tasks.json
Comment (xtophs) [inline containers/cpp-mariadb/.vscode/tasks.json:2]: not valid json ... hence the red syntax highlighting

**Review by xtophs — CHANGES_REQUESTED** [review]

File: containers/cpp-mariadb/test-project/main.cpp
Comment (xtophs) [inline containers/cpp-mariadb/test-project/main.cpp:15]: is this the right place for the `try`? It seems a little generous since getenv for sure doesn't throw. The line with `sql::mariadb::get_driver_instance` seems like the right place.

File: containers/cpp-mariadb/test-project/main.cpp
Comment (xtophs) [inline containers/cpp-mariadb/test-project/main.cpp:46]: Why not combine declaration and assignment to improve conscieseness and readability, i.e. 

`sql::Statement *stmt = conn->createStatement();`

File: containers/cpp-mariadb/test-project/main.cpp
Comment (xtophs) [inline containers/cpp-mariadb/test-project/main.cpp:33]: conn requires checking for null. See https://mariadb.com/docs/clients/mariadb-connectors/connector-cpp/connect/#sql-driver-connect. 

A null pointer would not result in a `sql::SQLException` and thus crash the program

File: containers/cpp-mariadb/test-project/main.cpp
Comment (xtophs) [inline containers/cpp-mariadb/test-project/main.cpp:46]: why not use smart pointers as done with `conn`? Mixing patterns is confusing.

File: containers/cpp-mariadb/test-project/main.cpp
Comment (xtophs) [inline containers/cpp-mariadb/test-project/main.cpp:58]: use of smart pointers would eliminate the need for those 2 lines.

File: containers/cpp-mariadb/test-project/main.cpp
Comment (xtophs) [inline containers/cpp-mariadb/test-project/main.cpp:60]: Did you mean to call `conn.close()` here? 

`release()` on releases the unique_ptr's ownership of the connection. If I read the docs correctly, you'd have to call `delete` on the returned connection pointer to free up the memory of the underlying object. 

https://en.cppreference.com/w/cpp/memory/unique_ptr/release

**Review by xtophs — CHANGES_REQUESTED** [review]

File: containers/cpp-mariadb/test-project/main.cpp
Comment (xtophs) [inline containers/cpp-mariadb/test-project/main.cpp:49]: It's not a cluster.

File: containers/cpp-mariadb/test-project/main.cpp
Comment (juzuluag) [reply containers/cpp-mariadb/test-project/main.cpp:15]: Great comments @xtophs. I'll take care of `main.cpp`

File: containers/cpp-mariadb/.devcontainer/install-mariadb.sh
Comment (KanishkT123) [reply containers/cpp-mariadb/.devcontainer/install-mariadb.sh:1]: We wrote this script as a whole.

File: containers/cpp-mariadb/.vscode/tasks.json
Comment (KanishkT123) [reply containers/cpp-mariadb/.vscode/tasks.json:2]: The invalid JSON is part of the original container. I don't think this repo makes a differentiation between JSON and JSON With Comments.

File: containers/cpp-mariadb/.devcontainer/docker-compose.yml
Comment (KanishkT123) [reply containers/cpp-mariadb/.devcontainer/docker-compose.yml]: Done!

File: containers/cpp-mariadb/test-project/main.cpp
Comment (dnastrain) [reply containers/cpp-mariadb/test-project/main.cpp]: hi @juzuluag, my understanding is xtophs recommended we move the `try` to right before the line `sql::Driver* driver = sql::mariadb::get_driver_instance`, rather than remove it altogether.

File: containers/cpp-mariadb/test-project/main.cpp
Comment (dnastrain) [reply containers/cpp-mariadb/test-project/main.cpp]: hi @juzuluag, my $0.02 is that here (and above for the `try`), that xtophs is suggesting we either:

**(a)** do what this link suggested and test the `conn` for null: https://mariadb.com/docs/clients/mariadb-connectors/connector-cpp/connect/#sql-driver-connect

**(b)** or, we do what this page suggested and  catch the `sql::SQLException`: https://mariadb.com/docs/clients/mariadb-connectors/connector-cpp/connect/#code-example-connect

File: containers/cpp-mariadb/.devcontainer/docker-compose.yml
Comment (juzuluag) [inline containers/cpp-mariadb/.devcontainer/docker-compose.yml:42]: @KanishkT123 , 'container_name' is not needed

File: containers/cpp-mariadb/.devcontainer/.env
Comment (juzuluag) [inline containers/cpp-mariadb/.devcontainer/.env:5]:
```suggestion
MARIADB_HOSTNAME=localhost
```

**Review by juzuluag — COMMENTED** [review]
@KanishkT123 , couple of suggested changes to be incorporated.

File: containers/cpp-mariadb/.devcontainer/docker-compose.yml
Comment (Chuxel) [inline containers/cpp-mariadb/.devcontainer/docker-compose.yml:35]: Same comment here as in the go definition.  We default the container itself to running as root, but then connect to it using non-root via the remoteUser property. (This is like doing `docker exec -u vscode` after the container has started).  You can then have the entrypoint fire system processes as needed while running as non-root and even uninstalling sudo if you're worried about it.

**Review by Chuxel — COMMENTED** [review]
Same comment as in the Go definition for consideration.

File: containers/cpp-mariadb/.devcontainer/docker-compose.yml
Comment (juzuluag) [reply containers/cpp-mariadb/.devcontainer/docker-compose.yml]: Hi @Chuxel. Fixed with the [latest](https://github.com/microsoft/vscode-dev-containers/pull/1241/commits/92e49336c0ebb4969efbdff5f9902e62a841ee2b) push

**Comment by Chuxel** [comment]
LGTM!

**Pull request merged**
